arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stream/callbacks.kt

```diff
@@ -0,0 +1,194 @@
+ *       end()
+ *     }
+ *     .compile()
+ *     .toList()
+ *     .let(::println) //[1, 2, 3, 4]
+ * //sampleEnd
+ * ```
+ *
+ * Note that if neither `end()` nor other limit operators such as `take(N)` are called,
+ * then the Stream will never end.
+ */
+@UseExperimental(ExperimentalTypeInference::class)
+fun <A> Stream.Companion.callback(@BuilderInference f: suspend EmitterSyntax<A>.() -> Unit): Stream<A> =
+  Stream.cancellable(f.andThen { CancelToken.unit })
+
+/**
+ * Creates a cancellable Stream from the given suspended block that will evaluate the passed [CancelToken] if cancelled.
+ *
+ * The suspending [f] runs in an uncancellable manner, acquiring [CancelToken] as a resource.
+ * If cancellation signal is received while [cb] is running, then the [CancelToken] will be triggered as soon as it's returned.
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.coroutines.CancelToken
+ * import arrow.fx.coroutines.stream.*
+ * import java.lang.RuntimeException
+ * import java.util.concurrent.Executors
+ * import java.util.concurrent.ScheduledFuture
+ * import java.util.concurrent.TimeUnit
+ *
+ * typealias Callback = (List<String>?, Throwable?) -> Unit
+ *
+ * class GithubId
+ * object GithubService {
+ *   private val listeners: MutableMap<GithubId, ScheduledFuture<*>> = mutableMapOf()
+ *   fun getUsernames(callback: Callback): GithubId {
+ *     val id = GithubId()
+ *     val future = Executors.newScheduledThreadPool(1).run {
+ *       var count = 0
+ *       scheduleAtFixedRate({
+ *         callback(listOf("Arrow - ${count++}"), null)
+ *       }, 0, 500, TimeUnit.MILLISECONDS)
+ *     }
+ *     listeners[id] = future
+ *     return id
+ *   }
+ *
+ *   fun unregisterCallback(id: GithubId): Unit {
+ *     listeners[id]?.cancel(false)
+ *     listeners.remove(id)
+ *   }
+ * }
+ *
+ * suspend fun main(): Unit {
+ *   //sampleStart
+ *   fun getUsernames(): Stream<String> =
+ *     Stream.cancellable {
+ *       val id = GithubService.getUsernames { names, throwable ->
+ *         when {
+ *           names != null -> emit(names)
+ *           throwable != null -> throw throwable
+ *           else -> throw RuntimeException("Null result and no exception")
+ *         }
+ *       }
+ *
+ *       CancelToken { GithubService.unregisterCallback(id) }
+ *     }
+ *
+ *   val result = getUsernames()
+ *     .take(3)
+ *     .compile()
+ *     .toList()
+ *
+ *   //sampleEnd
+ *   println(result)
+ * }
+ * ```
+ *
+ * If neither `end()` nor other limit operators such as `take(N)` are called,
+ * then the Stream will never end.
+ */
+@UseExperimental(ExperimentalTypeInference::class)
+fun <A> Stream.Companion.cancellable(@BuilderInference f: suspend EmitterSyntax<A>.() -> CancelToken): Stream<A> =
+  force {
+    val q = Queue.unbounded<Chunk<A>>()
+    val error = UnsafePromise<Throwable>()
+    val cancel = Promise<CancelToken>()
+
+    Stream.bracketCase({
+      ForkAndForget { emitterCallback(f, cancel, error, q) }
```

Comment:
@nomisRev is there a reason on why not using `ForkConnected` here instead?

Reply:
The acquire step of `bracket` runs in an `uncancellable` manner, so there is no cancellation to hook into.
So here instead we acquire the `Fiber` as a resource, and we `cancel` it when our interruptable stream finishes.